### PR TITLE
Bazelify & fix deadlock in record.Addr()

### PR DESCRIPTION
**Pull request by chrismcbride**

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -45,33 +43,39 @@ func Record(ttl time.Duration, addrs []string) *record {
...
 
-func (r *record) Expired() bool {
-	valid := time.Now().Before(r.expires)
-	if !valid {
-		go func() { r.stop <- true }()
-	}
-	return !valid
+// Returns a random address ignoring the record expiry.
+func (r *record) RandomAddr() string {
+	n := rand.Int() % len(r.addrs)
```
Comment (markchadwick) [inline cache.go:77]: or: n: = rand.Int31n(len(r.addrs))

File: cache.go [hunk 2]
```diff
@@ -88,43 +92,39 @@ func New(ttl time.Duration) *Cache {
...
 
+	// no cache and the lookup failed, just proxy the error
 	if err != nil {
 		lookupErr.Mark(1)
 		return "", err
 	}
 
-	rec := Record(c.ttl, hosts)
+	// store the new record
+	rec = Record(c.ttl, hosts)
```
Comment (markchadwick) [inline cache.go]: I can't seem to get the comment on :124 in this file. But, I think the write lock should go up 'round :105. It's not going to have a concurrent modification down her perse, but we will have a thundering herd after a cache expiration all doing the same work then lining up to write (I imagine) the same value

**Review by markchadwick — COMMENTED** [review]
Let me know what you think about moving the write lock

File: cache.go (hunk 2 above)
Comment (chrismcbride) [reply cache.go]: yea makes sense, ill take a look

Comment (chrismcbride) [reply cache.go]: suppose that would make this library less useful for caching many distinct hosts since thats a global lock

**Comment by chrismcbride** [comment]
Made new pull request with fixes because I did something stupid with git...forgot I wasnt using gerrit. Closing this one.

**Pull request closed without merging**